sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -184,6 +202,21 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
     case a @ Aggregate(_, _, p @ Project(_, Join(_, right, RightOuter, _, _)))
         if a.groupOnly && p.references.subsetOf(right.outputSet) =>
       a.copy(child = p.copy(child = right))
+
+    case a @ Aggregate(_, aggExprs, Join(left, _, LeftOuter, _, _))
+        if a.references.subsetOf(left.outputSet) && allDuplicateAgnostic(aggExprs) =>
+      a.copy(child = left)
+    case a @ Aggregate(_, aggExprs, Join(_, right, RightOuter, _, _))
+        if a.references.subsetOf(right.outputSet) && allDuplicateAgnostic(aggExprs) =>
+      a.copy(child = right)
+    case a @ Aggregate(_, aggExprs, p @ Project(projectList, Join(left, _, LeftOuter, _, _)))
```

Comment:
How about we make the optimizer rules more orthogonal and merge adjacent Aggregate and Project in `CollapseProject`

Reply:
Seems it's hrad. `CollapseProject` supports collapse `Project(_, Aggregate)` but does not support `Aggregate(_, _, Porject)` and that is we want to optimize.

The reason why we do not support is that we may lose column prune if we collapse `Aggregate(_, _, Porject)` without check the child of Porject.

Reply:
OK, but we should share the same concern in `CollapseProject`. This optimization may increase the number of times to evaluate the project list expressions, and we need to make sure they are cheap to evaluate.